akka-projection-grpc/src/main/scala/akka/persistence/EventWriter.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.LoggerOps
+import akka.actor.typed.scaladsl.adapter.TypedActorRefOps
+import akka.annotation.InternalApi
+import akka.pattern.StatusReply
+import akka.persistence.query.typed.EventEnvelope
+
+import java.util
+import java.util.UUID
+
+// FIXME move to akka-persistence-typed for access
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object EventWriter {
+
+  sealed trait Command
+  final case class Write(envelope: EventEnvelope[Any], replyTo: ActorRef[StatusReply[Done]]) extends Command
+
+  def apply(journalPluginId: String): Behavior[Command] =
+    Behaviors
+      .setup[AnyRef] { context =>
+        val writerUuid = UUID.randomUUID().toString
+        val journal = Persistence(context.system).journalFor(journalPluginId)
+        context.log.debug("Event writer for journal [{}] starting up", journalPluginId)
+
+        val waitingForResponse = new util.HashMap[(String, Long), ActorRef[StatusReply[Done]]]()
+
+        Behaviors.receiveMessage {
+          case Write(envelope, replyTo) =>
+            // FIXME trace or remove
+            context.log.debug(
+              "Writing event persistence id [{}], sequence nr [{}]",
+              envelope.persistenceId,
+              envelope.sequenceNr)
+            val repr = PersistentRepr(
+              envelope.event,
+              persistenceId = envelope.persistenceId,
+              sequenceNr = envelope.sequenceNr,
+              manifest = "", // adapters would be on the producing side, already applied
+              writerUuid = writerUuid,
+              sender = akka.actor.ActorRef.noSender)
+
+            val write = AtomicWrite(envelope.eventMetadata match {
+                case Some(meta) => repr.withMetadata(meta)
+                case _          => repr
+              }) :: Nil
+
+            waitingForResponse.put((envelope.persistenceId, envelope.sequenceNr), replyTo)
```

Comment:
We could add micro-batching here. If a new Write is received when there is already a pending response for that pid we would add it to a buffer for that pid instead of sending to journal immediately. Then we would flush that buffer when the response arrives, possibly batching more than one event into the same AtomicWrite.

Reply:
Sounds worth exploring

Reply:
Tried it out in [1b2d19a](https://github.com/akka/akka-projection/pull/932/commits/1b2d19a9651b697fe6079311109dd12813b3f7ac)

Reply:
Doesn't seem like that batching actually works, investigating.

Reply:
I had a 1s sleep between events in the producer 🤦

Reply:
Heh, also: https://github.com/akka/akka-persistence-r2dbc/pull/414